pyomo/common/multithread.py

```diff
@@ -0,0 +1,57 @@
+from collections import defaultdict
+from threading import get_ident, current_thread, main_thread
+
+
+class MultiThreadWrapper():
+    def __init__(self, base):
+        self.__mtdict = defaultdict(base)
+
+    def __getattr__(self, attr):
+        id = get_ident()
+        return getattr(self.__mtdict[id], attr)
+    
+    def __setattr__(self, attr, value):
+        if attr == '_MultiThreadWrapper__mtdict':
+            object.__setattr__(self, attr, value)
+        else:
+            id = get_ident()
+            setattr(self.__mtdict[id], attr, value)
```

Comment:
I would suggest reimplementing this as:
```python
def __getattr__(self, attr):
    if attr == 'main_thread':
        return self.__mtdict[main_thread().ident]
    else:
        id = get_ident()
        return getattr(self.__mtdict[id], attr)

def __setattr__(self, attr, value):
    if attr == '_MultiThreadWrapper__mtdict':
        object.__setattr__(self, attr, value)
    elif attr == 'main_thread':
        raise ValueError('Setting `main_thread` attribute is not allowed')
    else:
        id = get_ident()
        setattr(self.__mtdict[id], attr, value)
```

Reply:
The idea behind having two classes is that in some cases (i.e. PauseGC) the main thread instance reference is not needed (in fact I used `MultiThreadWrapper` instead of `MultiThreadWrapperWithMain`). I like the clear, almost self-explanatory way `MultiThreadWrapper` is implemented. MultiThreadWrapperWithMain is just a different way of treating the instance corresponding to the main_thread() id.

In general, I like to separate responsibilities of classes as a principle.

Nevertheless, I like the name `main_thread` instead of just main.

> I think that for the current implementation of `MultiThreadWrapperWithMain`, you are ending up with two instances of the wrapper class for the main thread.

How could it happen? I think this would be a fixable bug if so.

Reply:
I just thought of a simpler implementation of MultiThreadWrapperWithMain.. let me know what you think

Reply:
I'll try to write a simpler implementation of the wrapper with main

Reply:
Done. Please, review the newer one,